Added Support for Vector Extensions

shubhamnarlawar77 opened the pull request.

shubhamnarlawar77 committed:
Added Support for Vector Extensions

shubhamnarlawar77 commented:
@andikleen : review request - Is approach used for generating vector extension fine?

1. I have closed previous pull request due to small conflict occurred after adding Variable & Label attributes but have included same commit in this pull request. (I have taken initial notes suggested by Andi during start of Vector Extensions).

2. Mentioning @marxin in this pull request also.

marxin commented:
Ok, so first I would add to `VectorVariable` two member variables:
- `element_type`
- `num_elements`

Then you can have e.g. `ev32qu` which will have `VectorVariable(CHAR_TYPE,  32)`. Having that this way, you can easily do:

```			case ev32qu:    v = GenerateRandomCharConstant();		break;
```
which can be done based on `element_type`. Similarly for code like this:
```
	vector<string> vector_extension = {"int v48i", "unsigned int v8su", "short v16hi",\
```
These can be also automatically generated. About the names, you can make `typedef`s based on element type and number of elements as well.

marxin commented:
As another step, I would implement the same interface which we have for arrays. That is an assignment to an element of a vector type and getting element value. Doing that you'll have vector basics working. Next step can be then to support couple of `__builtin`s on the vector types, and the arithmetics on vector types, etc.

I'll be on vacation next week.

shubhamnarlawar77 commented:
> Ok, so first I would add to `VectorVariable` two member variables:
> 
>     * `element_type`
> 
>     * `num_elements`
> 
> 
> Then you can have e.g. `ev32qu` which will have `VectorVariable(CHAR_TYPE, 32)`. Having that this way, you can easily do:

Are you missing here something? Here, I can see blank code i.e. `     ` but no code in it. (after `you can easily do:`)

marxin commented:
It's about how you call all the 'GenerateRandomCharConstant()' and other functions in the switch statement.

marxin commented:
Is there are a progress of this pull request?

shubhamnarlawar77 commented:
> Is there are a progress of this pull request?

Yes. I will push changes soon!

shubhamnarlawar77 commented:
> Ok, so first I would add to `VectorVariable` two member variables:
> 
>     * `element_type`
> 
>     * `num_elements`
Added these 2 member variables.

Now, due to vector types pushed in `AllTypes`, pointer and array variables are also getting type as vector type. That means changes in type class may not exist if member functions in class `VectorVariable` are created similar to Array variable.

for example - `transparent_crc()` calculates checksum for normal variable and array variable but not vector variable. Hence, a member function in `VectorVariable` has to be added.

I will sent more changes in next push and brief comment also.

shubhamnarlawar77 commented:
Now, after trying several approaches on how can I generate vector variables along with normal & array variables with altering csmith grammer. I have used below approach to generate vector variables to make program compilable and let happen existing operations and arithmetics on vector variables also.

Approach - 
1. With the help of 2 member function `element_type` and `num_elements` suggested by Martin, I have introduced a new function `get_string_from_type()` which will generate alternate vector string based on data type of current type.

2. Fixed most of errors such as `transparent_crc` was not there for vector variable.

The current push is completely compilable if program generated with option `--no-global-variables`. There is only 1 error to fix i.e. selecting vector variable as assignment to normal variables. The next step is to fix this error.

Note - Changes in this push may be required after above error fix as I have used some hacks to make at least a compilable program containing vector extension.

@marxin Please check current push and suggest changes.

marxin commented on src/Constant.cpp:
```diff
@@ -385,7 +385,8 @@ GenerateRandomConstant(const Type* type)
 		}
-	} else {
+	}
+	else {
```
Please revert this.

shubhamnarlawar77 replied:
Done

marxin commented on src/DefaultOutputMgr.cpp:
```diff
@@ -206,6 +206,8 @@ DefaultOutputMgr::Output()
 	}
 	else {
+		if(CGOptions::vector_extension())
+			OutputVectorDeclarations(out);
```
Maybe make the decision in `CGOptions::vector_extension()` in `OutputVectorDeclarations`?

shubhamnarlawar77 replied:
Done.

marxin commented on src/Type.cpp:
```diff
@@ -452,6 +452,44 @@ Type::get_type_from_string(const string &type_string)
+string
+Type::get_string_from_type(const Type *type)
+{
+	if (type->simple_type == eInt) {
```
I don't like the switch. Rather I would prefer to generate that automatically. Based on `simple_type` and `num_elements` and mabe `signed` flag, you should be able to generate that without a switch.

shubhamnarlawar77 replied:
How about a function like `string VectorVariable::get_vector_type(type->simple_type)` that will return `vector_type` based on respective `simple_type` using a switch case ?

for example - 
                        `case eChar:`
                        `return "v32qi";`
                        `break;`

Csmith has used these type of functions in Type class for converting type to string and viceversa. The approach here we followed is to generate `vector_type` for given `simple_type`.

marxin replied:
No, you have these information:
- types - char, short, int, long
- you know size of these types: `sizeof(char) == 1`
- you are filling vectors with `32 bytes`

Having that you can make all cartesian combinations and generate all types. You don't have to obey `v32qi` names, but you can do something like:
`v32char`, `v16short`, `v16ushort`, ...

shubhamnarlawar77 replied:
I have removed `get_string_from_type()` function and instead generating types in `VectorVariable` class itself with help of above parameters.

marxin commented on src/Type.cpp:
```diff
@@ -1966,6 +2004,22 @@ OutputStructUnionDeclarations(std::ostream &out)
+OutputVectorDeclarations(std::ostream &out)
+{
+	output_comment_line(out, "--- VECTOR DECLARATIONS ---");
+	vector<string> vector_extension = {"int v8si", "unsigned int v8su", "short v16hi",\
```
Likewise here, I don't like hardcoded string constants.

shubhamnarlawar77 replied:
Here, I am using a `map<string, string>` to store corressponding types - like char and v32char. Is this fine?

marxin replied:
I still don't like it much, but that detail for now. Please carry on working on the operations on vector types.

marxin commented on src/Type.h:
```diff
@@ -70,8 +70,8 @@ enum eTypeDesc
-	eVoid,
-	eChar,
+	eVoid,//
+	eChar,//v32qi
```
I would remove this.

shubhamnarlawar77 replied:
Sure.

marxin commented on src/Variable.cpp:
```diff
@@ -468,7 +469,19 @@ Variable::Variable(const std::string &name, const Type *type, const Expression*
+}
+
+//constructor for vector variable
+Variable::Variable(const std::string &name, const Type *type, const Expression* init, const CVQualifiers* qfer, const bool isVector)
```
Do you really need an extra constructor? What about default value for the last argument `isVector`?

shubhamnarlawar77 replied:
No. You are right. I can make use of existing constructor with 4 parameters. (Will push changes in next push of this pull request).

marxin commented on src/Variable.cpp:
```diff
@@ -742,6 +755,9 @@ Variable::Output(std::ostream &out) const
 		out << "ACCESS_ONCE(" << get_actual_name() << ")";
 	}
+	else if (isVector){
+		out << get_actual_name() << "[0]";
```
Does it mean you always access the first element of a vector type? Why not a random element?

shubhamnarlawar77 replied:
Yes.
I tried doing `"[" << rnd_flipcoin(num_elements) << "]"` but I get segmentation fault - floating point exception (couldn't figure out why it is happening). But yes agree with you, correct approach is to select random element.

marxin replied:
Then it's probably an access that is out of array. Have you tried `valgrind` or sanitizers:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3d0f17e..a9ab258 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -121,7 +121,7 @@ add_definitions("-DHAVE_CONFIG_H")
 
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
     OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-long-long")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-long-long -fsanitize=address,undefined")
   set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3")
 endif()
 
```

shubhamnarlawar77 replied:
Got it after some debugging with sanitize option. My bad. The function `rnd_flipcoin()` does not accept 0 and hence led us to segmentation fault.

(Now preparing for push from all your above suggested changes)

shubhamnarlawar77 replied:
Here, I am able to use `"[" << rnd_flipcoin(num_elements) << "]"` in the place of `"[0]"` now.

Now, I am left with one error to fix i.e assigning vector elements addresses to pointer variables (which is of course not correct).

marxin reviewed (changes requested):
If I see correctly you have issues when you take an address of a vector element:

```
1.c:190:2529: note: (near initialization for ‘g_2284[2][9][7]’)
1.c:190:2529: error: initializer element is not computable at load time
1.c:190:2529: note: (near initialization for ‘g_2284[2][9][7]’)
1.c: In function ‘func_1’:
1.c:223:24: warning: initialization of ‘uint32_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘long long unsigned int *’ [-Wincompatible-pointer-types]
  223 |     uint32_t *l_2607 = &g_1951[0];
```

?

shubhamnarlawar77 commented:
> If I see correctly you have issues when you take an address of a vector element:
> 
> ```
> 1.c:190:2529: note: (near initialization for ‘g_2284[2][9][7]’)
> 1.c:190:2529: error: initializer element is not computable at load time
> 1.c:190:2529: note: (near initialization for ‘g_2284[2][9][7]’)
> 1.c: In function ‘func_1’:
> 1.c:223:24: warning: initialization of ‘uint32_t *’ {aka ‘unsigned int *’} from incompatible pointer type ‘long long unsigned int *’ [-Wincompatible-pointer-types]
>   223 |     uint32_t *l_2607 = &g_1951[0];
> ```
> 
> ?

Yes, but this issue occurs for only global variables and not local variables. You can see there is no such error like -
`error: initializer element is not computable at load time`
`volatile int32_t *g_72 = &g_73[0];` for local variables of functions.
(I am figuring out why it is happening like this?)

And I guess this is the main issue of this pull request.

marxin commented:
> `error: initializer element is not computable at load time`
> `volatile int32_t *g_72 = &g_73[0];` for local variables of functions.
> (I am figuring out why it is happening like this?)

It's probably because you can't take an address of an element of a vector variable.
